### ISSUE TITLE
KPI Request: Percent of Test Failures Identified by TestGrid

### ISSUE THREAD

**Issue by hemajv** [aicoe-aiops/ocp-ci-analysis#114]
As an OpenShift product manager, I would like to see the percent of tests that passed/failed updated weekly, so that I can track and measure the effectiveness and success of our testing process.

By measuring this metric, we can observe the trend of test failures over time and a decrease in the percent of failures over time (or increase in percent of test passes), should correlate to improved test efficiencies, enhanced testing process and error free releases. 

**How to collect metric**

In testgrid, we see that each test consists of multiple runs and the output of each run can broadly be categorized as "pass", "fail", "flaky". ([this document](https://github.com/aicoe-aiops/ocp-ci-analysis/blob/master/docs/cell_labels_in_testgrid.md) defines the meaning of all values in the testgrid data) The testgrid platform already uses a heuristic to identify passing and failing test runs, and flags them in their data set as the value `12` and `1` respectively. Therefore, it should be simple enough to calculate the percent of test failures and test pass each week by simply summing the total number of tests and the total number of failures/pass in our data set, grouped by week. 

**Acceptance Criteria** 

- [x] small notebook in `notebooks/data-sources/TestGrid/metrics/` that collects the `test pass/failure` metrics and stores it in ceph as a parquet.
- [x] small notebook in `notebooks/data-sources/TestGrid/metrics/` that collects the `test build pass/failure` metrics and stores it in ceph as a parquet.
- [x] small notebook in `notebooks/data-sources/TestGrid/metrics/` that collects the `test blocked/timed out` metrics and stores it in ceph as a parquet.

**Related Issues**
#76

**Comment by hemajv** [comment]
cc @oindrillac please do add anything else I may have missed

**Comment by oindrillac** [comment]
The metrics that would be visualized over time as an outcome this notebook are as follows:

- Total number of test cases
- Number of test cases passed
- Number of test cases failed
- Test pass %
- Test fail %

**Comment by hemajv** [comment]
@MichaelClifford since there are multiple metrics that can be associated with the test run statuses, can we consolidate all of them into one notebook instead of having separate notebooks for each?

**Comment by MichaelClifford** [comment]
@hemajv  are you referring to metrics like:
>>Total number of test cases
Number of test cases passed
Number of test cases failed
Test pass %
Test fail %

**Comment by hemajv** [comment]
yes correct, we were also hoping to include additional ones like:
- Number of test cases blocked
- Test blocked %
- Number of test cases with build failures/pass
- Test build failure/pass %

**Comment by MichaelClifford** [comment]
yes. I think that is fine. One thing to consider tho is that when processing the data as a pipeline in the future multiple notebooks can be run in parallel. So if the additional metrics you want to include here are really independent of each other it might be better to have them as separate notebooks.

**Comment by oindrillac** [comment]
@MichaelClifford we followed the same procedure as outlined in the notebook above to look for `build pass/fails` and `blocked and timed-out` tests in separate notebooks as defined in our 2nd and 3rd acceptance criteria. 

However, we found that the count and percent of all the above metrics come out to be 0.

On further inspection, we tried deriving the count and percent or each label highlighted [here](https://github.com/aicoe-aiops/ocp-ci-analysis/blob/master/docs/cell_labels_in_testgrid.md). And we found that all come out to be 0 except 1, 4, 12, 13. We think that this is interesting and we can discuss this in our upcoming meeting.

cc: @hemajv

**Comment by MichaelClifford** [comment]
@oindrillac that sounds good, we can discuss in upcoming meeting :smiley:  . I wonder if this list of labels is newer than the sample dataset we are using `testgrid_810.json.gz`. Did you try to rerun `notebooks/data-sources/TestGrid/testgrid_EDA.ipynb` to generate a dataset based on today's testgrid values and re-run your notebooks?

**Comment by oindrillac** [comment]
> @oindrillac that sounds good, we can discuss in upcoming meeting  . I wonder if this list of labels is newer than the sample dataset we are using `testgrid_810.json.gz`. Did you try to rerun `notebooks/data-sources/TestGrid/testgrid_EDA.ipynb` to generate a dataset based on today's testgrid values and re-run your notebooks?

@MichaelClifford we can try this out on the newer data and see how that goes!

**Comment by oindrillac** [comment]
@MichaelClifford I re -an the analysis on the new dataset for today and I still get 0 for all other [labels](https://github.com/aicoe-aiops/ocp-ci-analysis/blob/master/docs/cell_labels_in_testgrid.md) except 0,1, 4, 12, 13.

**Comment by MichaelClifford** [comment]
@oindrillac ok, well then I guess it sounds like the values either are currently not being tracked or haven't occurred in the data. But having a notebook that can capture them if/ when they eventually occur is fine. 

Which is to say, if the work is already done, and the results are "No data for this metric currently" then lets add the notebooks and move on. If the notebooks are not ready to be included, let's leave them for now and move on. WDYT?

**Comment by oindrillac** [comment]
@MichaelClifford that makes sense. :+1: Already have notebooks for build pass/failure and blocked/timed out. Can add them to the repo.

**Comment by MichaelClifford** [comment]
@oindrillac I was just taking a closer look at the testgrid data, and you and @hemajv should confirm this, But I _think_ these "build_pass/fail"  values are not actually being used at all. However, they are capturing build success or failure by attributing it to the test "overall" if this is the case, you could update the your notebook to look at just the "overal" test pass/fail rate. WDYT?

**Comment by MichaelClifford** [comment]
You can confirm the last comment  by clicking on an "overall" test in testgrid, going into its prow page and seeing if it also includes a "pass" or "fail" value at the top.

**Comment by oindrillac** [comment]
@MichaelClifford we are trying to verify this by exporting a sample dataset from a different job name where the overall status is is "Passing" at times, to check if the build pass/fail values are consistent with that. Will update on what we see.

**Comment by oindrillac** [comment]
I tried running the build pass/fail notebook for the specific
**dashboard name** - "redhat-openshift-ocp-release-4.2-informing"
**job** - promote-release-openshift-machine-os-content-e2e-aws-4.2-s390x
**date** - 01/29/2021 

Upon inspecting the test "overall" and looking at the prow page, the value seems to "Passed"

![image](https://user-images.githubusercontent.com/32435206/110163050-8fd00e00-7dbd-11eb-8811-df3c0f41750b.png)

However, the build failure and build [passing values still seem to only hold "False" labels

![image](https://user-images.githubusercontent.com/32435206/110163575-33b9b980-7dbe-11eb-9a6b-cf19e227d66d.png)
![image](https://user-images.githubusercontent.com/32435206/110163646-4df39780-7dbe-11eb-91cc-8a208cbbf1aa.png)


And the cumulative value comes out to be "0" still 
![image](https://user-images.githubusercontent.com/32435206/110163335-e6d5e300-7dbd-11eb-9bc0-cf2e18d4b499.png)

cc: @MichaelClifford @hemajv

**Comment by hemajv** [comment]
@oindrillac thanks for digging into this! there definitely seems to be some mismatch, either the `build_pass/build_fail` columns are not actively being used or there is some other reasoning (which we are not aware of) as to how they are being defined

**Comment by MichaelClifford** [comment]
@oindrillac does the above show that the "build pass" and "build failure" flags aren't being used, but the test "overall" corresponds to the "passing" or "failing" label for builds in Prow, like we suspected? 

In that case we can use "overall" as our proxy for any of the build metrics we looking at?

**Comment by oindrillac** [comment]
> @oindrillac does the above show that the "build pass" and "build failure" flags aren't being used, but the test "overall" corresponds to the "passing" or "failing" label for builds in Prow, like we suspected?

@MichaelClifford yes that is true. The build pass/fail metrics don't seem to be used at all. However, IIUC according to Testgrid terminology, a collection of tests for a timestamp within a Job could be referred to as a Build, as it has a unique Build ID. This can be verified by hovering over the individual blocks along a vertical line within a job. Each test along a vertical grid has a unique build ID such as `1369630643398381568` which can also be found on the Prow page for the Build.

For example, along the vertical line, in the following 2 screenshots, as I hover over multiple tests corresponding to the same timestamp, they consist of the same highlighted build ID `1369630643398381568`.

![build-id](https://user-images.githubusercontent.com/32435206/110716688-9ef7f700-81d5-11eb-8f12-053c84961aba.png)

![build-id-2](https://user-images.githubusercontent.com/32435206/110716696-a3241480-81d5-11eb-8d63-4ceb371d0491.png)

For each build, the `overall` seems to be a Flag which contains a Pass/Fail value based on an aggregate of the Tests within the Build.

> In that case we can use "overall" as our proxy for any of the build metrics we looking at?

So, therefore, the "overall" value in Prow, can be good proxy for "build" pass/fail.

**Comment by MichaelClifford** [comment]
@oindrillac great! Thanks for digging into this. So, lets use the "overall" test address any questions about the "build"
